src/vs/workbench/services/textfile/electron-browser/textFileService.ts

```diff
@@ -55,20 +56,73 @@ export class TextFileService extends AbstractTextFileService {
 	}
 
 	public resolveTextContent(resource: URI, options?: IResolveContentOptions): TPromise<IRawTextContent> {
-		return this.fileService.resolveStreamContent(resource, options).then(streamContent => {
-			return ModelBuilder.fromStringStream(streamContent.value).then(res => {
-				const r: IRawTextContent = {
-					resource: streamContent.resource,
-					name: streamContent.name,
-					mtime: streamContent.mtime,
-					etag: streamContent.etag,
-					encoding: streamContent.encoding,
-					value: res.value,
-					valueLogicalHash: res.hash
-				};
-				return r;
+
+		const result: IRawTextContent = {
+			resource: undefined,
+			name: undefined,
+			mtime: undefined,
+			etag: undefined,
+			encoding: undefined,
+			value: undefined,
+			valueLogicalHash: undefined,
+		};
+
+		let contentToken = new CancellationTokenSource();
+		let completePromise: Thenable<any>;
+
+		let statsPromise = this.fileService.resolveFile(resource).then(stat => {
+			result.resource = stat.resource;
+			result.name = stat.name;
+			result.mtime = stat.mtime;
+			result.etag = stat.etag;
+
+			if (stat.size > MAX_FILE_SIZE) {
+				// stop reading the file. the stat and content resolve call
+				// usually race, mostly likely the stat call will win and cancel
+				// the content call
+				contentToken.cancel();
+				throw new FileOperationError(
+					nls.localize('fileTooLargeError', "File too large to open"),
+					FileOperationResult.FILE_TOO_LARGE
+				);
+			}
+		});
+
+		if (options && options.etag) {
+			// await the stat iff we already have an etag so that we compare the
+			// etag from the stat before we actually read the file again.
+			completePromise = statsPromise.then(() => {
+				// Return early if file not modified since
+				if (result.etag === options.etag) {
+					return TPromise.wrapError<IRawTextContent>(new FileOperationError(
+						nls.localize('fileNotModifiedError', "File not modified since"),
+						FileOperationResult.FILE_NOT_MODIFIED_SINCE)
+					);
+				}
+				// Waterfall -> only now resolve the contents
+				return this.fillInContents(result, resource, options, contentToken.token);
 			});
+
+		} else {
+			// a fresh load without a previous etag which means we can resolve the file stat
+			// and the content at the same time, avoiding the waterfall.
+			completePromise = Promise.all([statsPromise, this.fillInContents(result, resource, options, contentToken.token)]);
+		}
+
+		return TPromise.wrap(completePromise).then(() => result);
+	}
+
+	private fillInContents(content: IRawTextContent, resource: URI, options: IResolveContentOptions, token: CancellationToken): Thenable<any> {
```

Comment:
What is the reason for not doing all of this from within the file service? We used to have a method `resolveStreamContent` that was used for the editor model loading that contains both stream and metadata and I think we can preserve all of that also with your solution. Imho the code here should be pushed down and we can just keep `resolveStreamContent` with your faster implementation?

Reply:
> What is the reason for not doing all of this from within the file service?

No-one uses the file service like that, only the text file service... My understanding is that every other caller of `resolveContent` and `resolveContentStream` cares only about the actual data and the encoding. So, why make everyone pay for this? I'd prefer to see the file service as something more raw which the text file service uses to compose something more complex.

Reply:
I do not have a good argument other than a) making it harder to test this change if parts of it is spread between `fileService` and `textFileService` and b) making it harder to understand how the method behaves when called from the editor. I do think that something like the NOT_MODIFIED_SINCE is something that a remote file system couldn't implement as well (e.g. that was actually taken from the HTTP 304 Not Modified as inspiration).